### PR TITLE
scripts: check requirements

### PR DESCRIPTION
**Pull request by overvenus**
#843 

OS X is not tested.

### CODE REVIEW

**Comment by ngaut** [comment]
Please do not use absolute path. Please see https://docs.minio.io/docs/how-to-install-golang

**Comment by ngaut** [comment]
Ping @overvenus

**Comment by overvenus** [comment]
@ngaut 
I come across a problem, check_req.sh forks from make, so `export` will not affect other rules in Makefile which means they can't find go even if it has been installed. I am working on it.

**Comment by overvenus** [comment]
PTAL @siddontang @ngaut

File: scripts/check_req.sh [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                ;;
+            esac
+        ;;
+
+        darwin*)
+            # refer to https://github.com/facebook/rocksdb/blob/master/INSTALL.md
+            xcode-select --install
+            brew update
+            brew tap homebrew/versions
+            brew install gcc47 --use-llvm
```
Comment (siddontang) [inline scripts/check_req.sh]: can we use gcc48 here? you use > 4.7 in linux.

File: scripts/check_req.sh [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                ;;
+                CentOS)
+                    if [ ! -d $GOROOT ]; then
+                        curl -L https://storage.googleapis.com/golang/go1.6.2.linux-amd64.tar.gz -o golang.tar.gz
+                        tar -C ${DEPS_PATH}/ -xzf golang.tar.gz
+                        ${SUDO} yum install -y hg git
+                    fi
+                ;;
+                *)
+                    echo "unsupported platform $dist, you may install RocksDB manually"
```
Comment (siddontang) [inline scripts/check_req.sh]: not rocksdb

File: scripts/check_req.sh [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+                    # refer to https://github.com/golang/go/wiki/Ubuntu
+                    ${SUDO} add-apt-repository ppa:ubuntu-lxc/lxd-stable
+                    ${SUDO} apt-get update
+                    ${SUDO} apt-get install -y golang mercurial git
+                ;;
+                CentOS)
+                    if [ ! -d $GOROOT ]; then
+                        curl -L https://storage.googleapis.com/golang/go1.6.2.linux-amd64.tar.gz -o golang.tar.gz
+                        tar -C ${DEPS_PATH}/ -xzf golang.tar.gz
+                        ${SUDO} yum install -y hg git
```
Comment (siddontang) [inline scripts/check_req.sh]: why do we need install hg and git here?

File: scripts/check_req.sh [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+function install_go {
+    echo "Intall go ..."
+    case "$OSTYPE" in 
+        linux*) 
+            dist=$(get_linux_platform)
+            case $dist in
+                Ubuntu)
+                    # refer to https://github.com/golang/go/wiki/Ubuntu
```
Comment (siddontang) [inline scripts/check_req.sh]: I think we can use linux package for all linux platforms.

File: scripts/check_req.sh (hunk 3 above)
Comment (overvenus) [inline scripts/check_req.sh]: `go get` needs git and hg

File: scripts/check_req.sh (hunk 4 above)
Comment (overvenus) [inline scripts/check_req.sh]: Do you mean install go in `tmp`?

Comment (siddontang) [inline scripts/check_req.sh]: yes, or we can install in common /usr/local/go and set PATH in bashrc too.

File: scripts/check_req.sh [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        *)
+            echo "unsupported $OSTYPE"
+            exit 1
+        ;;
+    esac
+}
+
+# Check rust
+if ! which cargo > /dev/null; then
+    curl -sSf https://static.rust-lang.org/rustup.sh | sh -s -- --channel=nightly
```
Comment (siddontang) [inline scripts/check_req.sh]: in my previous centos test, I must use sudo here, otherwise, I can't install rust...

Comment (overvenus) [inline scripts/check_req.sh]: My bad, I skip testing rust because of slow network.

File: scripts/check_req.sh [hunk 6]
```diff
@@ -101,8 +101,13 @@ function install_gpp {
 }
 
 # Check rust
-if ! which cargo > /dev/null; then
-    curl -sSf https://static.rust-lang.org/rustup.sh | sh -s -- --channel=nightly
+if which cargo > /dev/null; then
+    if ! cargo --version | grep nightly > /dev/null; then
+        echo "Please upgrade Rust to nightly."
```
Comment (siddontang) [inline scripts/check_req.sh]: should exit

File: scripts/check_req.sh [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            dist=$(get_linux_platform)
+            case $dist in
+                Ubuntu)
+                    ${SUDO} apt-get install -y g++
+                ;;
+                CentOS)
+                    ${SUDO} yum install -y gcc-c++
+                ;;
+                *)
+                    echo "unsupported platform $dist, you may install RocksDB manually"
```
Comment (siddontang) [inline scripts/check_req.sh]: not rocksdb

**Comment by siddontang** [comment]
LGTM

File: Makefile [hunk 8]
```diff
@@ -3,6 +3,10 @@ ENABLE_FEATURES ?= default
 DEPS_PATH = $(CURDIR)/tmp
 BIN_PATH = $(CURDIR)/bin
 
+ifeq "$(GOROOT)" ""
```
Comment (BusyJay) [inline Makefile]: GOROOT ?= $(DEPS_PATH)/go

**Comment by siddontang** [comment]
PTAL @BusyJay

File: scripts/check_req.sh [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+        *)
+            echo "unsupported $OSTYPE"
+            exit 1
+        ;;
+    esac
+}
+
+# Check rust
+if which cargo > /dev/null; then
```
Comment (BusyJay) [inline scripts/check_req.sh]: &>/dev/null

File: scripts/check_req.sh [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        exit 1
+    fi
+else
+    install_go
+fi
+
+# Check g++
+if which g++ > /dev/null; then
+    # Check g++ version, RocksDB requires >= 4.7
+    G_VER_1=`g++ --version | awk 'match($0, /([0-9])+(\.[0-9])+/) { ver = substr($0, RSTART, RLENGTH) ; ver = split(ver, n, ".") ; print n[1]}'`
```
Comment (BusyJay) [inline scripts/check_req.sh]: g++ -dumpversion

**Comment by BusyJay** [comment]
rest LGTM

File: Makefile [hunk 11]
```diff
@@ -3,6 +3,7 @@ ENABLE_FEATURES ?= default
 DEPS_PATH = $(CURDIR)/tmp
 BIN_PATH = $(CURDIR)/bin
 GOROOT ?= $(DEPS_PATH)/go
+export GOROOT
```
Comment (siddontang) [inline Makefile]: why export?

Comment (overvenus) [inline Makefile]: In CentOS, script_req can not find GOROOT if we dont export.

Comment (siddontang) [inline Makefile]: pass this to script_req too.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**